### PR TITLE
Favorites

### LINKED ISSUE #20: Favorites View
Add a way to select a GIF as a favorite to store it. the view will already be declared after #18

### LINKED ISSUE #26: Investigate how to see Gif in the views list (history and favs)

### PR DESCRIPTION
**Pull request by Carglass**
Ticket: fixes #20, fixes #26 

## Changes

- Created a Favorites view:
  - click the star in history to add a Gif to favorites
  - one can insert a Gif from the favorites
  - one can remove a Gif from the favorites
- Clicking the item name in one of the views now displays this item in a webview

## How To Test

- build in debug
- reset your history (to get the view from clicking on one element, the click the title command is set in the constructor of the item, old items will not have it)
- do a search to add something to the history
- add this gif to favorites
- insert this gif from the favorites
- click on the item name in both views to check that the gif is displayed
- remove the item from the favorites

### CODE REVIEW

File: src/addToFavorites.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	state: vscode.Memento,
+	favoritesTreeView: FavoritesProvider
+) => void = (
+	gif: HistoryEntry,
+	state: vscode.Memento,
+	favoritesTreeView: FavoritesProvider
+) => {
+	const favorites: FavoritesEntry[] | undefined = state.get('favorites');
+	if (favorites) {
+		if (favorites.filter(entry => entry.gifUri === gif.gifUri).length !== 0) {
```
Comment (fallanic) [inline src/addToFavorites.ts]: we can simplify this
```suggestion
		if (favorites.find(entry => entry.gifUri === gif.gifUri)) {
```

File: src/addToFavorites.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			vscode.window.showInformationMessage(
+				'This GIF was already in your favorites'
+			);
+		} else {
+			const newFav = new FavoritesEntry(gif.label, gif.gifUri);
+			state.update('favorites', [newFav].concat(favorites)).then(() => {
```
Comment (fallanic) [inline src/addToFavorites.ts]: no limit for favorites? Not saying we should, just wondering if it's a design choice or a missing statement. I'd rather have unlimited favorites if it works

File: src/favorites.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+		return element;
+	}
+
+	getChildren(element?: FavoritesEntry): Thenable<FavoritesEntry[]> {
+		if (element) {
+			// our tree has only one level
+			return Promise.resolve([]);
+		} else {
+			// from the root, returns the history list
+			return Promise.resolve(this.favorites ? this.favorites : []);
```
Comment (fallanic) [inline src/favorites.ts]: Just a suggestion to keep the code lean (the same applies to the history one).

```suggestion
			return Promise.resolve(this.favorites || []);
```

File: src/history.ts [hunk 4]
```diff
@@ -40,9 +40,9 @@ export class HistoryEntry extends vscode.TreeItem {
 		super(label);
 		// this command is the one when clicking the name, could be removed, or changed to viewing the gif?
 		this.command = {
-			command: 'giflens.addGif',
-			title: 'Giflens Add Gif',
-			arguments: [gifUri],
+			command: 'giflens.viewGif',
```
Comment (fallanic) [inline src/history.ts]: at first I was like "is that a mistake? 🤔" and then "OMG he actually implemented it 😄"

File: src/addToFavorites.ts (hunk 2 above)
Comment (Carglass) [reply src/addToFavorites.ts]: yes, I thought that favorites were user controlled, while the history keeps getting bigger

File: src/removeFromFavorites.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		const updatedFavorites = favorites.filter(
+			entry => entry.gifUri !== gif.gifUri
+		);
+		if (updatedFavorites.length < favorites.length) {
+			state.update('favorites', updatedFavorites).then(() => {
+				favoritesTreeView.refresh(state);
+			});
+		} else {
+			throw new Error(
+				'remove from Favorites has been called from a Gif not present in the favorites'
```
[the remark this answers is not in the document]
Comment (Carglass) [reply src/removeFromFavorites.ts]: ben pas trop mais un if sans else, c'est mieux de l'éviter non ?

[the remark this answers is not in the document]
Comment (Carglass) [reply src/removeFromFavorites.ts]: le if est nécessaire pour typescript, puisque concrètement, on ne sait pas ce qu'il y a dans le globalState

[the remark this answers is not in the document]
Comment (fallanic) [reply src/removeFromFavorites.ts]: 🇫🇷

File: src/viewGif.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	 */
+	init(
+		gif: HistoryEntry | FavoritesEntry,
+		editor: vscode.TextEditor | undefined
+	) {
+		if (editor) {
+			this.editor = editor;
+		} else {
+			vscode.window.showInformationMessage(
+				'You have no active editor, you wont be able to click-insert the Gif'
```
Comment (fallanic) [inline src/viewGif.ts]: nice 👍

**Comment by Carglass** [comment]
> Works great 👏
> 
> Side note : after doing a search, I notice we cannot redo a search if the result panel is still open.
> Something to look at after this release.
> 
> :shipit:

created #30 for this issue

**Pull request merged**